### PR TITLE
Added RequiredGrad attribute to YieldOp

### PR DESCRIPTION
**Pull request by satyajandhyala**
**Description**: Added RequiredGrad attribute to YieldOp

**Motivation and Context**
- Certain outputs from the input model do not required gradients
- If it fixes an open issue, please link to the issue here.
https://msdata.visualstudio.com/Vienna/_workitems/edit/1031167
- Fixes the issue taking incorrect gradient output from PyTorch

### CODE REVIEW

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 1]
```diff
@@ -2220,13 +2220,23 @@ Return true if all elements are true and false otherwise.
       .Output(0, "outputs_grad", "Gradient of outputs returned from pytorch.", "T", OpSchema::Variadic,
               /*is_homogeneous*/ false,
               /*min_arity*/ 1)
+      .Attr(
+        "RequiredGrad",
+        "The number of inputs with require_glad flag set.",
+        AttributeProto::INT,
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2226]: INTS

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.cc [hunk 2]
```diff
@@ -188,42 +191,42 @@ void ModuleGradientGraphBuilder::AddYieldOp() {
...
+  // YieldOps RequiredGrad attribute specifies the indices of the required gradients.
+  ONNX_NAMESPACE::AttributeProto required_grad;
+  const std::string attribute_name = "RequiredGrad";
+  required_grad.set_name(attribute_name);
+  required_grad.set_type(ONNX_NAMESPACE::AttributeProto::INTS);
+
   // Yield inputs include all user outputs, those require output gradients come first, so Yield Op can use their shapes
   // to infer Op output shapes.
   std::vector<std::string> user_output_names_require_grad;
   std::vector<std::string> user_output_names_no_grad;
```
Comment (iK1D) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.cc:203]: These two are no longer needed. We can remove it and also the emplace_back call for these two vectors below.

**Review by iK1D — APPROVED** [review]

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.cc [hunk 3]
```diff
@@ -50,7 +50,10 @@ Status ModuleGradientGraphBuilder::Initialize(std::istream& model_istream,
 
   // Remove the training initializers from the graph and move them to graph inputs.
   for (const auto& initializer_name : training_graph_info_.initializer_names_to_train) {
-    input_args.emplace_back(graph.GetNodeArg(initializer_name));
+    const NodeArg* node_arg = graph.GetNodeArg(initializer_name);
+    if (node_arg) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.cc:54]: maybe replace the if  check with ORT_ENFORCE(node_arg != nullptr) ?

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 4]
```diff
@@ -2220,19 +2220,27 @@ Return true if all elements are true and false otherwise.
       .Output(0, "outputs_grad", "Gradient of outputs returned from pytorch.", "T", OpSchema::Variadic,
               /*is_homogeneous*/ false,
               /*min_arity*/ 1)
+      .Attr(
+        "RequiredGrad",
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2224]: nit. the naming convention for attribute is "required_grad"

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 5]
```diff
@@ -2220,19 +2220,27 @@ Return true if all elements are true and false otherwise.
...
+        "The indices of required gradient outputs.",
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2225]: nit. "The input indices that required gradients."

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.h [hunk 6]
```diff
@@ -41,8 +41,8 @@ struct TrainingGraphInfo {
   std::vector<std::string> initializer_grad_names_to_train{};
   // The user outputs.
   std::vector<std::string> user_output_names{};
-  // The user output grad names that are actual required by the backward graph.
-  std::vector<std::string> backward_output_grad_names{};
+  // The user output grad names that are actual required by the backward graph mapped to the index of the correspoinding output of inference graph.
+  std::map<std::string, size_t> backward_output_grad_names_map{};
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/module_gradient_graph_builder.h:45]: unordered_map

File: orttraining/orttraining/python/training/ortmodule.py [hunk 7]
```diff
@@ -414,14 +414,15 @@ def forward(ctx, *inputs, **kwargs):
...
-            def backward(ctx, *grad_output):
+            def backward(ctx, *grad_outputs):
                 '''Performs backward pass based on grad wrt module output
                 '''
 
                 # Use IO binding
                 # Push user output grads to ONNX backend.
                 backward_grad_output_ortvalue = []
-                for grad_output in grad_output[:len(self._onnx_graphs_info.backward_output_grad_names)]:
+                for _, i in self._onnx_graphs_info.backward_output_grad_names_map.items():
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:424]: Let's add a comment here 
backward_output_grad_names_map only contains the subset of module outputs that need a gradient, we filter out the invalid entries in 'grad_outputs'

File: samples/python/mnist/graph_spliter.py [hunk 8]
```diff
@@ -85,7 +85,7 @@ def print_type(name, type):
 print_list('backward_intializer_names_as_input', split_graphs_info.backward_intializer_names_as_input)
 print_list('intermediate_tensor_names', split_graphs_info.intermediate_tensor_names)
 print_list('user_output_grad_names', split_graphs_info.user_output_grad_names)
-print_list('backward_output_grad_names', split_graphs_info.backward_output_grad_names)
+print_list('backward_output_grad_names_map', split_graphs_info.backward_output_grad_names_map)
```
Comment (SherlockNoMad) [inline samples/python/mnist/graph_spliter.py:88]: this file is outdated and should actually be removed. 
Let's keep this file untouched, and we can removed it in another PR.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 5 above)
Comment (satyajandhyala) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2225]: I think it may be better if I change it to "The indices of the required gradient outputs."

File: samples/python/mnist/graph_spliter.py (hunk 8 above)
Comment (satyajandhyala) [reply samples/python/mnist/graph_spliter.py:88]: I guess we should keep the edit for consistency.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 9]
```diff
@@ -357,6 +389,67 @@ def test_input_requires_grad_backward_creates_input_grad(device):
...
+    y1, y2 = model(x1, x2)
+    s = y2.sum()
+    s.backward()
+    assert x1.grad is not None and x2.grad is not None
+
+
+@pytest.mark.parametrize("device", ['cuda', 'cpu'])
+def test_input_requires_grad_backward_creates_input_grad_as_required1(device):
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetMultiplePositionalArgumentsMultipleOutputs(D_in, H, D_out).to(device)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:408]: These four tests can be combined into one with pytest.mark.parametrize 
since the only changing variable is requires_grad for x1 and x2

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 10]
```diff
@@ -357,6 +389,67 @@ def test_input_requires_grad_backward_creates_input_grad(device):
...
+    model = ORTModule(model)
+    x1 = torch.randn(N, D_in, device=device, requires_grad=True)
+    x2 = torch.randn(N, D_in, device=device, requires_grad=False)
+    y1, y2 = model(x1, x2)
+    s = y2.sum()
+    s.backward()
+    assert x1.grad is not None and x2.grad is None
+
+@pytest.mark.parametrize("device", ['cuda', 'cpu'])
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:417]: with only 'cuda' is probably good enough

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 11]
```diff
@@ -357,6 +389,67 @@ def test_input_requires_grad_backward_creates_input_grad(device):
...
+def test_input_requires_grad_backward_creates_input_grad_as_required0(device):
+    N, D_in, H, D_out = 32, 784, 500, 10
+    model = NeuralNetMultiplePositionalArgumentsMultipleOutputs2(D_in, H, D_out).to(device)
+    model = ORTModule(model)
+    x1 = torch.randn(N, D_in, device=device, requires_grad=True)
+    x2 = torch.randn(N, D_in, device=device, requires_grad=True)
+    y1, y2 = model(x1, x2)
+    s = y2.sum()
+    s.backward()
+    assert x1.grad is not None and x2.grad is not None
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:402]: a more interesting assert would be 
the param.grad for the y1 branch should be zero.

File: orttraining/orttraining/core/framework/module_gradient_graph_builder.cc (hunk 3 above)
Comment (satyajandhyala) [reply orttraining/orttraining/core/framework/module_gradient_graph_builder.cc:54]: If we have an unused layer such as the following defined in the module __init__ function the corresponding node_arg is zero at this point. 
        self.fc2 = torch.nn.Linear(hidden_size, num_classes)
Should we still keep the assertion?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 11 above)
Comment (satyajandhyala) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:402]: at this stage, y1.grad and y2.grad both are null because the results only contain the gradients of the inputs with require_grad = True. The grad_outputs are not added to results in ortmodule.py. We could add the assertion that unrequired gradient outputs are zero in ortmodule itself.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 12]
```diff
@@ -519,6 +549,31 @@ def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder
...
+    N, D_in, H, D_out = 32, 784, 500, 10
+    device = 'cuda'
+    model = NeuralNetMultiplePositionalArgumentsMultipleOutputs1(D_in, H, D_out).to(device)
+    model = ORTModule(model)
+    x1 = torch.randn(N, D_in, device=device, requires_grad=x1_requires_grad)
+    x2 = torch.randn(N, D_in, device=device, requires_grad=x2_requires_grad)
+    y1, y2 = model(x1, x2)
+    s = y2.sum()
+    s.backward()
+    assert (not x1_requires_grad or x1.grad is not None) and (not x2_requires_grad or x2.grad is not None)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:575]: can we add the asserts on the x1.grad and x2.grad correctness? 

see test_multiple_overlapping_forward_backward_calls for example

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 13]
```diff
@@ -550,6 +582,71 @@ def test_changes_input_requires_grad_reinitializes_module_gradient_graph_builder
...
+
+    def run_step(model, x1, x2):
+        y1, y2 = model(x1, x2)
+        s = y2.sum()
+        s.backward()
+
+    N, D_in, H, D_out = 32, 784, 500, 10
+    device = 'cuda'
+    pt_model = NeuralNetMultiplePositionalArgumentsMultipleOutputs1(D_in, H, D_out).to(device)
+    ort_model = ORTModule(pt_model)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:632]: nit. 
we should create a deepcopy of the pt_model here, otherwise ort_model and pt_model are sharing the same underlaying weights. 

it's ok for this test case, because we are not updating the weight values. If we have another test case that modify the weight values in run_step(), then you will see unmatching results.

**Review by SherlockNoMad — APPROVED** [review]
